packages/collaboration-manager/src/Operation.ts

```diff
@@ -56,4 +56,130 @@ export class Operation {
     this.index = index;
     this.data = data;
   }
+
+  /**
+   * Makes an inverse operation
+   */
+  public inverse(): Operation {
+    const index = this.index;
+
+    switch (this.type) {
+      case OperationType.Insert:
+
+        const textRange = index.textRange;
+
+        if (textRange == undefined) {
+          throw new Error('Unsupported index');
+        }
+
+        const [ textRangeStart ] = textRange;
+
+        const newIndex = new IndexBuilder()
+          .from(index)
+          .addTextRange([textRangeStart, textRangeStart + this.data.newValue.length])
+          .build();
+
+        return new Operation(OperationType.Delete, newIndex, {
+          prevValue: this.data.newValue,
+          newValue: this.data.prevValue,
+        });
+      case OperationType.Delete:
+        return new Operation(OperationType.Insert, index, {
+          prevValue: this.data.newValue,
+          newValue: this.data.prevValue,
+        });
+      case OperationType.Modify:
+        return new Operation(OperationType.Modify, index, {
+          prevValue: this.data.newValue,
+          newValue: this.data.prevValue,
+        });
+    }
+  }
+
+  /**
+   * Transforms the operation against another operation
+   *
+   * @param againstOp - operation to transform against
+   */
+  public transform(againstOp: Operation): Operation {
+    if (!this.index.isTextIndex || !againstOp.index.isTextIndex) {
+      throw new Error('Unsupported index');
+    }
+
+    if (this.type === OperationType.Modify || againstOp.type === OperationType.Modify) {
+      throw new Error('Unsupported operation type');
+    }
+
+    /**
+     * Do not transform operations if they are on different blocks or documents
+     */
+    if (this.index.documentId !== againstOp.index.documentId || this.index.blockIndex !== againstOp.index.blockIndex) {
```

Comment:
We need to consider block index for text operations, if block before was added or deleted, we need to update  the index

Reply:
Can be done in a separate PR